Fixes #30556 - Get an accurate err for invalid hosts in report

shiramax opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

shiramax committed:
Fixes #30556 - Get an accurate err for invalid hosts in report

tbrisker commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -17,39 +17,47 @@ template_inputs:
+  <%-     status = "#{host2.name} only"-%>
+  <%-   elsif version2 == "-" -%>
+  <%-     status = "#{host1.name} only"-%>
+  <%-   elsif version(version1) > version(version2) -%>
```
I think you need to rebase, these should be `sortable_version`

tbrisker commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -17,39 +17,47 @@ template_inputs:
 model: ReportTemplate
```
While you're at it, can you add
```
require:
- plugin: katello
  version: 3.16.1
```
to the metadata so it is only seeded when using katello with a version that supports it?

ares commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -15,41 +15,54 @@ template_inputs:
+  -%>
+<%- else -%>
+
+  <%- host1 = hosts.first.first -%>
```
please don't indent the `<%`, use inner indentation like below, e.g.
```
<%-   host1 = hosts.first.first -%>
```

that way the code is properly indented inside of ERB tags

ares commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -15,41 +15,54 @@ template_inputs:
--%>
+
+<%- hosts = load_hosts(search: "name ^ (#{input('Host 1')}, #{input('Host 2')})") -%>
+<%- if hosts.first.to_a.length !=2 -%>
```
missing space before `2`

ares commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -15,41 +15,54 @@ template_inputs:
+<%- hosts = load_hosts(search: "name ^ (#{input('Host 1')}, #{input('Host 2')})") -%>
+<%- if hosts.first.to_a.length !=2 -%>
+  <%- report_row(
+    'Error': "One of the hosts:#{input('Host 1')}, #{input('Host 2')} is not valid "
```
this will render report normally with 200 as response, however we should clearly indicate that rendering has failed

I think we should raise an error instead and define a specific exception for this case at https://github.com/theforeman/foreman/blob/develop/lib/foreman/renderer/errors.rb

The message should say why the host is invalid. Perhaps turn it into "At least one of the hosts couldn't be found" or something like that.

tbrisker commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -15,41 +15,51 @@ template_inputs:
-        'Status': status
-      )
+
+<%- hosts = load_hosts(search: "name ^ (#{input('Host 1')}, #{input('Host 2')})") -%>
+<%- if hosts.first.to_a.length != 2 -%>
```
This returns an `Enumerator` from the `in_batches`, you can use
```suggestion
<%- hosts = load_hosts(search: "name ^ (#{input('Host 1')}, #{input('Host 2')})").to_a.flatten -%>
<%- if hosts.length != 2 -%>
```
to get rid of all the `.first` calls and make the `hosts` object a bit easier to work with

tbrisker commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -15,41 +15,51 @@ template_inputs:
+<%- else -%>
+
+<%-   host1 = hosts.first.first -%>
+<%-   host2 = hosts.first.first -%>
```
This is probably wrong, assuming you take the other suggestion it would be 
```suggestion
<%-   host2 = hosts.last -%>
```

tbrisker reviewed (dismissed):
lgtm, @ares any more comments?

tbrisker commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -15,41 +15,51 @@ template_inputs:
   resource_type: Katello::ActivationKey
 model: ReportTemplate
+require:
+- plugin: katello
+  version: 3.16.1
```
We should add a migration to drop this template for users who don't have katello installed, otherwise it won't be updated when seeding (and in any case this template shouldn't be there for them).
You can check for something like
``` Foreman::Plugin.find('katello') ```

tbrisker replied:
added the migration for you.

tbrisker committed:
Refs #30556 - Drop compare content hosts template

The template should only be present in Katello. Removing it from
non-katello installations.

ares commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -15,41 +15,51 @@ template_inputs:
-        'Architecture': arch_list[0],
-        'Status': status
-      )
+
```
this will caus an empty line in resulting report, breaking the format (e.g. json/yaml/csv)

ares commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -15,41 +15,51 @@ template_inputs:
+<%- if hosts.length != 2 -%>
+<%-   raise _('At least one of the hosts couldn\'t be found') -%>
+<%- else -%>
+
```
extra empty line, this would break the result

ares reviewed (changes requested):
Before I test, there are 2 extra empty lines that should be removed, please ping me when you update it

tbrisker committed:
Refs #30556 - fix spacing

ares commented on app/views/unattended/report_templates/host_-_compare_content_hosts_packages.erb:
```diff
@@ -15,41 +15,48 @@ template_inputs:
 model: ReportTemplate
+require:
+- plugin: katello
+  version: 3.16.1
```
does it really require 3.16.1? it seems to be working with 3.16.0 too, I'm not sure if 3.16.1 is even available already - https://fedorapeople.org/groups/katello/releases/yum/3.16/katello/el7/x86_64/

tbrisker replied:
I think that `sortable_version` will only be available in 3.16.1 - https://projects.theforeman.org/issues/30561

ares reviewed (changes requested):
one more thing, otherwise works as expected

ares commented:
Thanks @shiramax and @tbrisker, finally merged!

tbrisker commented:
2.2 - 7ce9201b3

tbrisker commented:
2.1 - 474b83d65
